<h1 align=center><font size = 6>Segmenting and Clustering</font></h1>

<h1 align=center><font size = 12>Neighborhoods in Toronto</font></h1>

<h1 align=center><font size = 10>.*.</font></h1>

<h1 align=center><font size = 5>*......***..................* ...*****.............**......***..................*</font></h1>

# Table of Contents


### Question 1: 

#### 1.1. Create a new Notebook 

#### 1.2. Scrape Web Page

#### 1.3. Transform Data into Pandas Dataframe 

#### 1.4. Clean Dataframe and Annotate Notebook

### 1.5. Link Notebook on Github repository. (10 marks)

https://github.com/lindangulopez/Coursera_Capstone/blob/master/Peer-graded%20Assignment%20Q1_Toronto_Linda.ipynb


### Question 2: 

#### 2.1. Get the latitude and the longitude coordinates of each neighborhood, used Geopy:

https://geopy.readthedocs.io/en/stable/#installation

#### 2.2. Use Geocoder Python package: https://geocoder.readthedocs.io/index.html. 

#### 2.3. Use Geopy & OpenStreetMap to create Dataframe

#### 2.4. Retrieved only 5 coordinates with with lambda equation, 

#### did not create a loop for each postal code to remake the calls.

#### 2.5. Used link to a csv file, http://cocl.us/Geospatial_data to create Dataframe:

https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html 

### 2.6. Submit a link to your Notebook on your Github repository. (2 marks)


<h1 align=center><font size = 8>...</font></h1>


<h1 align=center><font size = 5>*......***..................* ...*****.............**......***..................*</font></h1>



# Question 1:


## 1.1.  Create a new Notebook 

with the basic dependencies.

In [1]:
import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis
import requests # Library for web scraping

print('Libraries imported.')

Libraries imported.


## 1.2. Scrape Web Page

About the Data, Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, 
- is a list of postal codes in Canada where the first letter is M. Postal codes beginning with M are located within the city of Toronto in the province of Ontario. 

- Scraping table from HTML using BeautifulSoup, write a Python program similar to scrape.py,from:

### Corey Schafer Python Programming Tutorial:
The code from this video can be found at: https://github.com/CoreyMSchafer/code...




In [2]:
# To run this, you can install BeautifulSoup
# https://pypi.python.org/pypi/beautifulsoup4

# Or download the file
# http://beautiful-soup-4
# and unzip it in the same directory as this file
import requests
from urllib.request import urlopen
from bs4 import BeautifulSoup
import ssl
import csv

print('BeautifulSoup  & csv imported.')

BeautifulSoup  & csv imported.


In [3]:
# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

print('SSL certificate errors ignored.')

SSL certificate errors ignored.


In [4]:
source = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

soup = BeautifulSoup(source, 'lxml')

#print(soup.prettify())
print('soup ready')

soup ready


Now we Need to search within 'table', class_="wikitable sortable"table, each of the entry for the 'Postcode', 'Borough' and 'Neighbourhood' columns:

#### Typical tags within:

- (tbody & table) are,

- tr, td, and a 

#### Typical data are as follows, for 'Postcode', 'Borough' and 'Neighbourhood' respectively: 

- M8Z

- href="/wiki/Etobicoke" title="Etobicoke"

- South of Bloor

or 

- M9Z

- Not assigned

- Not assigned



In [5]:
table = soup.find('table',{'class':'wikitable sortable'})
#table

In [6]:

table_rows = table.find_all('tr')

#table_rows

In [7]:
data = []
for row in table_rows:
    data.append([t.text.strip() for t in row.find_all('td')])

df = pd.DataFrame(data, columns=['PostalCode', 'Borough', 'Neighbourhood'])
df = df[~df['PostalCode'].isnull()]  # to filter out bad rows

print(df.head(5))
print('***')
print(df.tail(5))

  PostalCode           Borough     Neighbourhood
1        M1A      Not assigned      Not assigned
2        M2A      Not assigned      Not assigned
3        M3A        North York         Parkwoods
4        M4A        North York  Victoria Village
5        M5A  Downtown Toronto      Harbourfront
***
    PostalCode       Borough          Neighbourhood
284        M8Z     Etobicoke              Mimico NW
285        M8Z     Etobicoke     The Queensway West
286        M8Z     Etobicoke  Royal York South West
287        M8Z     Etobicoke         South of Bloor
288        M9Z  Not assigned           Not assigned


## 1.3. Transform Data into Pandas Dataframe 

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 288 entries, 1 to 288
Data columns (total 3 columns):
PostalCode       288 non-null object
Borough          288 non-null object
Neighbourhood    288 non-null object
dtypes: object(3)
memory usage: 9.0+ KB


In [9]:
df.shape

(288, 3)

## 1.4. Clean Dataframe and Annotate Notebook

Only process the cells that have an assigned borough, we can ignore cells with 'Not assigned' boroughs, like in rows 1 & 2.

In [10]:
import pandas
import requests
from bs4 import BeautifulSoup
website_text = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(website_text,'lxml')

table = soup.find('table',{'class':'wikitable sortable'})
table_rows = table.find_all('tr')

data = []
for row in table_rows:
    data.append([t.text.strip() for t in row.find_all('td')])

df = pandas.DataFrame(data, columns=['PostalCode', 'Borough', 'Neighbourhood'])
df = df[~df['PostalCode'].isnull()]  # to filter out bad rows

df.head(15)

PostalCode           Borough     Neighbourhood
1         M1A      Not assigned      Not assigned
2         M2A      Not assigned      Not assigned
3         M3A        North York         Parkwoods
4         M4A        North York  Victoria Village
5         M5A  Downtown Toronto      Harbourfront
6         M5A  Downtown Toronto       Regent Park
7         M6A        North York  Lawrence Heights
8         M6A        North York    Lawrence Manor
9         M7A      Queen's Park      Not assigned
10        M8A      Not assigned      Not assigned
11        M9A         Etobicoke  Islington Avenue
12        M1B       Scarborough             Rouge
13        M1B       Scarborough           Malvern
14        M2B      Not assigned      Not assigned
15        M3B        North York   Don Mills North

In [11]:

df.drop(df[df['Borough']=="Not assigned"].index,axis=0, inplace=True)
df.head()

PostalCode           Borough     Neighbourhood
3        M3A        North York         Parkwoods
4        M4A        North York  Victoria Village
5        M5A  Downtown Toronto      Harbourfront
6        M5A  Downtown Toronto       Regent Park
7        M6A        North York  Lawrence Heights

The dataframe can be reindex as follows:

In [12]:
df1 = df.reset_index()
df1.head()

index PostalCode           Borough     Neighbourhood
0      3        M3A        North York         Parkwoods
1      4        M4A        North York  Victoria Village
2      5        M5A  Downtown Toronto      Harbourfront
3      6        M5A  Downtown Toronto       Regent Park
4      7        M6A        North York  Lawrence Heights

In [13]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211 entries, 0 to 210
Data columns (total 4 columns):
index            211 non-null int64
PostalCode       211 non-null object
Borough          211 non-null object
Neighbourhood    211 non-null object
dtypes: int64(1), object(3)
memory usage: 6.7+ KB


In [14]:
df1.shape

(211, 4)

More than one neighborhood can exist in one postal code area, M5A is listed twice and has two neighborhoods Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma using groupby, see: 

https://pandas-docs.github.io/pandas-docs-travis/user_guide/groupby.html


In [15]:
df2= df1.groupby('PostalCode').agg(lambda x: ','.join(x))

df2.head()
       

Borough  \
PostalCode                                        
M1B                     Scarborough,Scarborough   
M1C         Scarborough,Scarborough,Scarborough   
M1E         Scarborough,Scarborough,Scarborough   
M1G                                 Scarborough   
M1H                                 Scarborough   

                                   Neighbourhood  
PostalCode                                        
M1B                                Rouge,Malvern  
M1C         Highland Creek,Rouge Hill,Port Union  
M1E              Guildwood,Morningside,West Hill  
M1G                                       Woburn  
M1H                                    Cedarbrae

In [16]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 103 entries, M1B to M9W
Data columns (total 2 columns):
Borough          103 non-null object
Neighbourhood    103 non-null object
dtypes: object(2)
memory usage: 2.4+ KB


In [17]:
df2.shape

(103, 2)

There are also cells that have an assigned neighbouhoods,like M7A, lets assign their boroughs as their neighbourhood, as follows:

In [18]:

df2.loc[df2['Neighbourhood']=="Not assigned",'Neighbourhood']=df2.loc[df2['Neighbourhood']=="Not assigned",'Borough']

df2.head()



Borough  \
PostalCode                                        
M1B                     Scarborough,Scarborough   
M1C         Scarborough,Scarborough,Scarborough   
M1E         Scarborough,Scarborough,Scarborough   
M1G                                 Scarborough   
M1H                                 Scarborough   

                                   Neighbourhood  
PostalCode                                        
M1B                                Rouge,Malvern  
M1C         Highland Creek,Rouge Hill,Port Union  
M1E              Guildwood,Morningside,West Hill  
M1G                                       Woburn  
M1H                                    Cedarbrae

In [19]:
df3 = df2.reset_index()
df3.head()

PostalCode                              Borough  \
0        M1B              Scarborough,Scarborough   
1        M1C  Scarborough,Scarborough,Scarborough   
2        M1E  Scarborough,Scarborough,Scarborough   
3        M1G                          Scarborough   
4        M1H                          Scarborough   

                          Neighbourhood  
0                         Rouge,Malvern  
1  Highland Creek,Rouge Hill,Port Union  
2       Guildwood,Morningside,West Hill  
3                                Woburn  
4                             Cedarbrae

Now we can remove the duplicate boroughts as follows:

In [20]:
df3['Borough']= df3['Borough'].str.replace('nan|[{}\s]','').str.split(',').apply(set).str.join(',').str.strip(',').str.replace(",{2,}",",")

In [21]:
df3.head()

PostalCode      Borough                         Neighbourhood
0        M1B  Scarborough                         Rouge,Malvern
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Guildwood,Morningside,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

In [22]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103 entries, 0 to 102
Data columns (total 3 columns):
PostalCode       103 non-null object
Borough          103 non-null object
Neighbourhood    103 non-null object
dtypes: object(3)
memory usage: 2.5+ KB


In [23]:
df3.shape

(103, 3)

## 1.5 Dataframe cleaned & shared on Github:



https://github.com/lindangulopez/Coursera_Capstone/blob/master/Peer-graded%20Assignment%20Q1_Toronto_Linda.ipynb


# Question 2: 



## 2.1. Get the latitude and the longitude coordinates of each neighborhood, used Geopy:

https://geopy.readthedocs.io/en/stable/#installation


In [24]:
pip install geopy

Note: you may need to restart the kernel to use updated packages.


In [25]:
from  geopy.geocoders import Nominatim
geolocator = Nominatim()
city ="London"
country ="Uk"
loc = geolocator.geocode(city+','+ country)
print("latitude is :-" ,loc.latitude,"\nlongtitude is:-" ,loc.longitude)

latitude is :- 51.5073219 
longtitude is:- -0.1276474


In [26]:
from  geopy.geocoders import Nominatim
geolocator = Nominatim()
location = geolocator.geocode("Toronto, North York, Parkwoods")
print(location.address)

print((location.latitude, location.longitude))

print(location.raw)


Parkwoods Village Drive, Parkway East, Don Valley East, North York, Toronto, Ontario, M3A 1Z5, Canada
(43.7612239, -79.3239857)
{'place_id': 113868274, 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright', 'osm_type': 'way', 'osm_id': 160406963, 'boundingbox': ['43.761106', '43.761307', '-79.3242157', '-79.3239088'], 'lat': '43.7612239', 'lon': '-79.3239857', 'display_name': 'Parkwoods Village Drive, Parkway East, Don Valley East, North York, Toronto, Ontario, M3A 1Z5, Canada', 'class': 'highway', 'type': 'secondary', 'importance': 0.51}


In [27]:
import pandas as pd
df3.head()

PostalCode      Borough                         Neighbourhood
0        M1B  Scarborough                         Rouge,Malvern
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Guildwood,Morningside,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

In [28]:

import pandas as pd
df_geopy = pd.DataFrame({'PostalCode': ['M3A', 'M4A', 'M5A'],
                         'Borough': ['North York', 'North York', 'Downtown Toronto'],
                         'Neighbourhood': ['Parkwoods', 'Victoria Village', 'Harbourfront'],})

from geopy.geocoders import Nominatim
geolocator = Nominatim()

df_geopy['location'] = df_geopy[['PostalCode', 'Borough', 'Neighbourhood']].apply(lambda x: ', '.join(x), axis=1 )
df_geopy


PostalCode           Borough     Neighbourhood  \
0        M3A        North York         Parkwoods   
1        M4A        North York  Victoria Village   
2        M5A  Downtown Toronto      Harbourfront   

                              location  
0           M3A, North York, Parkwoods  
1    M4A, North York, Victoria Village  
2  M5A, Downtown Toronto, Harbourfront

In [29]:
df_geopy1 = df3

In [30]:
from geopy.geocoders import Nominatim
geolocator = Nominatim()

df_geopy1['location'] = df3[['PostalCode', 'Borough', 'Neighbourhood']].apply(lambda x: ', '.join(x), axis=1 )
df_geopy1.head()

PostalCode      Borough                         Neighbourhood  \
0        M1B  Scarborough                         Rouge,Malvern   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough       Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                Woburn   
4        M1H  Scarborough                             Cedarbrae   

                                            location  
0                    M1B, Scarborough, Rouge,Malvern  
1  M1C, Scarborough, Highland Creek,Rouge Hill,Po...  
2  M1E, Scarborough, Guildwood,Morningside,West Hill  
3                           M1G, Scarborough, Woburn  
4                        M1H, Scarborough, Cedarbrae

In [31]:
df_geopy1 = df3

In [32]:
df_geopy1.shape


(103, 4)

In [33]:
df_geopy1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103 entries, 0 to 102
Data columns (total 4 columns):
PostalCode       103 non-null object
Borough          103 non-null object
Neighbourhood    103 non-null object
location         103 non-null object
dtypes: object(4)
memory usage: 3.3+ KB


In [34]:
df_geopy1.drop(df_geopy1[df_geopy1['Borough']=="Notassigned"].index,axis=0, inplace=True)
df_geopy1
# code holds true up until i=102
df_geopy1.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 103 entries, 0 to 102
Data columns (total 4 columns):
PostalCode       103 non-null object
Borough          103 non-null object
Neighbourhood    103 non-null object
location         103 non-null object
dtypes: object(4)
memory usage: 4.0+ KB


In [35]:
df_geopy1.head()


PostalCode      Borough                         Neighbourhood  \
0        M1B  Scarborough                         Rouge,Malvern   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough       Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                Woburn   
4        M1H  Scarborough                             Cedarbrae   

                                            location  
0                    M1B, Scarborough, Rouge,Malvern  
1  M1C, Scarborough, Highland Creek,Rouge Hill,Po...  
2  M1E, Scarborough, Guildwood,Morningside,West Hill  
3                           M1G, Scarborough, Woburn  
4                        M1H, Scarborough, Cedarbrae

In [36]:
df_geopy1.shape


(103, 4)

In [37]:
df_geopy1.to_csv('geopy1.csv')
# no data for location after row 75

Now let's test for location = 'M1G, Scarborough, Woburn'

In [38]:
from  geopy.geocoders import Nominatim
geolocator = Nominatim()
location = geolocator.geocode("M1G, Scarborough, Woburn")
print(location.address)

print((location.latitude, location.longitude))

print(location.raw)


Woburn, Scarborough—Guildwood, Scarborough, Toronto, Golden Horseshoe, Ontario, M1H 2A2, Canada
(43.7598243, -79.2252908)
{'place_id': 4972888, 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright', 'osm_type': 'node', 'osm_id': 558715617, 'boundingbox': ['43.7597743', '43.7598743', '-79.2253408', '-79.2252408'], 'lat': '43.7598243', 'lon': '-79.2252908', 'display_name': 'Woburn, Scarborough—Guildwood, Scarborough, Toronto, Golden Horseshoe, Ontario, M1H 2A2, Canada', 'class': 'place', 'type': 'neighbourhood', 'importance': 0.582352868045614}


### Great getting latitude & longitude, with geopy



## 2.2. Use Geocoder Python package: https://geocoder.readthedocs.io/index.html. 


In [39]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.



Taking postal code M5G as an example, your code would look something like this:


> import geocoder # import geocoder

initialize your variables to None

> lat_lng_coords = None

> postal_code = None

loop until you get the coordinates

> while(lat_lng_coords is None):

>  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))

>  lat_lng_coords = g.latlng

> latitude = lat_lng_coords[0]

> longitude = lat_lng_coords[1]


print (latitude, longitude)

### Problem with this location / package / code kernel blocked :(


## 2.3. Use Geopy & OpenStreetMap to create Dataframe


In [40]:
df3.to_csv('geopy.csv')

In [41]:
import csv

with open('geopy.csv') as csvfile:
     reader = csv.DictReader(csvfile)
     for row in reader:
         print(row['PostalCode'],row['Borough'], row['Neighbourhood'] )

M1B Scarborough Rouge,Malvern
M1C Scarborough Highland Creek,Rouge Hill,Port Union
M1E Scarborough Guildwood,Morningside,West Hill
M1G Scarborough Woburn
M1H Scarborough Cedarbrae
M1J Scarborough Scarborough Village
M1K Scarborough East Birchmount Park,Ionview,Kennedy Park
M1L Scarborough Clairlea,Golden Mile,Oakridge
M1M Scarborough Cliffcrest,Cliffside,Scarborough Village West
M1N Scarborough Birch Cliff,Cliffside West
M1P Scarborough Dorset Park,Scarborough Town Centre,Wexford Heights
M1R Scarborough Maryvale,Wexford
M1S Scarborough Agincourt
M1T Scarborough Clarks Corners,Sullivan,Tam O'Shanter
M1V Scarborough Agincourt North,L'Amoreaux East,Milliken,Steeles East
M1W Scarborough L'Amoreaux West
M1X Scarborough Upper Rouge
M2H NorthYork Hillcrest Village
M2J NorthYork Fairview,Henry Farm,Oriole
M2K NorthYork Bayview Village
M2L NorthYork Silver Hills,York Mills
M2M NorthYork Newtonbrook,Willowdale
M2N NorthYork Willowdale South
M2P NorthYork York Mills West
M2R NorthYork Willowdale 

In [42]:
from  geopy.geocoders import Nominatim
geolocator = Nominatim()
location = geolocator.geocode("M1B Scarborough Rouge,Malvern")
print(location.address)

print((location.latitude, location.longitude))

print(location.raw)


Malvern, Scarborough—Rouge Park, Scarborough, Toronto, Golden Horseshoe, Ontario, M1B 4Y7, Canada
(43.8091955, -79.2217008)
{'place_id': 4979177, 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright', 'osm_type': 'node', 'osm_id': 558715616, 'boundingbox': ['43.8091455', '43.8092455', '-79.2217508', '-79.2216508'], 'lat': '43.8091955', 'lon': '-79.2217008', 'display_name': 'Malvern, Scarborough—Rouge Park, Scarborough, Toronto, Golden Horseshoe, Ontario, M1B 4Y7, Canada', 'class': 'place', 'type': 'neighbourhood', 'importance': 0.7914299961151839}


In [43]:
from  geopy.geocoders import Nominatim
geolocator = Nominatim()
location = geolocator.geocode("Toronto, Highland Creek")
print(location.address)

print((location.latitude, location.longitude))

print(location.raw)
#M1C Scarborough Highland Creek,Rouge Hill,Port Union = no address

Highland Creek, Scarborough—Rouge Park, Scarborough, Toronto, Golden Horseshoe, Ontario, M1C 3T7, Canada
(43.7901172, -79.1733344)
{'place_id': 332533, 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright', 'osm_type': 'node', 'osm_id': 139354747, 'boundingbox': ['43.7900672', '43.7901672', '-79.1733844', '-79.1732844'], 'lat': '43.7901172', 'lon': '-79.1733344', 'display_name': 'Highland Creek, Scarborough—Rouge Park, Scarborough, Toronto, Golden Horseshoe, Ontario, M1C 3T7, Canada', 'class': 'place', 'type': 'neighbourhood', 'importance': 0.6726559169652601}


In [44]:
from  geopy.geocoders import Nominatim
geolocator = Nominatim()
location = geolocator.geocode("Toronto, Morningside")
print(location.address)

print((location.latitude, location.longitude))

print(location.raw)

#M1E Scarborough Guildwood,Morningside,West Hill = no address


Morningside, Scarborough—Guildwood, Scarborough, Toronto, Ontario, M1G 3K5, Canada
(43.7826012, -79.2049579)
{'place_id': 332629, 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright', 'osm_type': 'node', 'osm_id': 139347757, 'boundingbox': ['43.7825512', '43.7826512', '-79.2050079', '-79.2049079'], 'lat': '43.7826012', 'lon': '-79.2049579', 'display_name': 'Morningside, Scarborough—Guildwood, Scarborough, Toronto, Ontario, M1G 3K5, Canada', 'class': 'place', 'type': 'neighbourhood', 'importance': 0.579877925155736}


### Geopy cannot find some full addresses, so let's just use the Borough name & Toronto to get the co_ordinates. 

Woburn, Scarborough—Guildwood, Scarborough, Toronto, Golden Horseshoe, Ontario, M1H 2A2, Canada
(43.7598243, -79.2252908)
{'place_id': 4972888, 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright', 'osm_type': 'node', 'osm_id': 558715617, 'boundingbox': ['43.7597743', '43.7598743', '-79.2253408', '-79.2252408'], 'lat': '43.7598243', 'lon': '-79.2252908', 'display_name': 'Woburn, Scarborough—Guildwood, Scarborough, Toronto, Golden Horseshoe, Ontario, M1H 2A2, Canada', 'class': 'place', 'type': 'neighbourhood', 'importance': 0.582352868045614}

Malvern, Scarborough—Rouge Park, Scarborough, Toronto, Golden Horseshoe, Ontario, M1B 4Y7, Canada
(43.8091955, -79.2217008)
{'place_id': 4979177, 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright', 'osm_type': 'node', 'osm_id': 558715616, 'boundingbox': ['43.8091455', '43.8092455', '-79.2217508', '-79.2216508'], 'lat': '43.8091955', 'lon': '-79.2217008', 'display_name': 'Malvern, Scarborough—Rouge Park, Scarborough, Toronto, Golden Horseshoe, Ontario, M1B 4Y7, Canada', 'class': 'place', 'type': 'neighbourhood', 'importance': 0.7914299961151839}

Highland Creek, Scarborough—Rouge Park, Scarborough, Toronto, Golden Horseshoe, Ontario, M1C 3T7, Canada
(43.7901172, -79.1733344)
{'place_id': 332533, 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright', 'osm_type': 'node', 'osm_id': 139354747, 'boundingbox': ['43.7900672', '43.7901672', '-79.1733844', '-79.1732844'], 'lat': '43.7901172', 'lon': '-79.1733344', 'display_name': 'Highland Creek, Scarborough—Rouge Park, Scarborough, Toronto, Golden Horseshoe, Ontario, M1C 3T7, Canada', 'class': 'place', 'type': 'neighbourhood', 'importance': 0.6726559169652601}


Morningside, Scarborough—Guildwood, Scarborough, Toronto, Ontario, M1G 3K5, Canada
(43.7826012, -79.2049579)
{'place_id': 332629, 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright', 'osm_type': 'node', 'osm_id': 139347757, 'boundingbox': ['43.7825512', '43.7826512', '-79.2050079', '-79.2049079'], 'lat': '43.7826012', 'lon': '-79.2049579', 'display_name': 'Morningside, Scarborough—Guildwood, Scarborough, Toronto, Ontario, M1G 3K5, Canada', 'class': 'place', 'type': 'neighbourhood', 'importance': 0.579877925155736}


In [45]:
import numpy as np
import csv


PostalCode = None
Borough = None
Neighbourhood = None
latData = None
longData = None

LAT_Woburn = 43.7598243
LONG_Woburn = -79.2252908
LAT_Malvern = 43.8091955
LONG_Malvern = -79.2217008
LAT_Highland_Creek = 43.7901172
LONG_Highland_Creek = -79.1733344
LAT_Morningside = 43.7826012
LONG_Morningside = -79.2049579

 
PostalCode = np.array(['M1H','M1B','M1C','M1G '])
Borough = np.array(['Scarborough','Scarborough','Scarborough','Scarborough'])
Neighbourhood = np.array(['Woburn','Malvern','Highland_Creek','Morningside'])
latData = np.array([43.7598243,43.8091955, 43.7901172 , 43.7826012])
longData = np.array([-79.2252908,-79.2217008,-79.1733344, -79.2049579 ])

with open('data.csv', 'w') as file:
    writer = csv.writer(file, delimiter=',')
    writer.writerow('ABXYZ')
    for a,b,x,y,z in np.nditer([ PostalCode.T, Borough.T, Neighbourhood.T, latData.T, longData.T], order='C'):
        writer.writerow([a,b,x,y,z])    

In [46]:
import csv

with open('data.csv') as csvfile:
     reader = csv.DictReader(csvfile)
     for row in reader:
         print(row['A'], row['B'],row['X'], row['Y'], row['Z'])

M1H Scarborough Woburn 43.7598243 -79.2252908
M1B Scarborough Malvern 43.8091955 -79.2217008
M1C Scarborough Highland_Creek 43.7901172 -79.1733344
M1G  Scarborough Morningside 43.7826012 -79.2049579


In [47]:
pd.read_csv('data.csv') 

A            B               X          Y          Z
0   M1H  Scarborough          Woburn  43.759824 -79.225291
1   M1B  Scarborough         Malvern  43.809196 -79.221701
2   M1C  Scarborough  Highland_Creek  43.790117 -79.173334
3  M1G   Scarborough     Morningside  43.782601 -79.204958


## 2.4. Retrieved only 5 coordinates with with lambda equation, 

#### did not create a loop for each postal code to remake the calls.



### Read the docs:

https://docs.python.org/3.5/library/fileformats.html

https://geopy.readthedocs.io/en/latest/index.html#geopy.geocoders.Nominatim

#### Watch: 44 Python Tutorial Example Geocoding Addresses with Pandas and Geopy

https://www.youtube.com/watch?v=q_OUHA_zqeM



### Load DataFrame

In [48]:
import pandas, os
#os.listdir()

In [49]:
df_geopy=df3
df_geopy.head()

PostalCode      Borough                         Neighbourhood  \
0        M1B  Scarborough                         Rouge,Malvern   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough       Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                Woburn   
4        M1H  Scarborough                             Cedarbrae   

                                            location  
0                    M1B, Scarborough, Rouge,Malvern  
1  M1C, Scarborough, Highland Creek,Rouge Hill,Po...  
2  M1E, Scarborough, Guildwood,Morningside,West Hill  
3                           M1G, Scarborough, Woburn  
4                        M1H, Scarborough, Cedarbrae

In [50]:
import geopy
dir(geopy)

['ArcGIS',
 'Baidu',
 'Bing',
 'DataBC',
 'GeoNames',
 'GeocodeFarm',
 'GeocoderDotUS',
 'GoogleV3',
 'IGNFrance',
 'LiveAddress',
 'Location',
 'NaviData',
 'Nominatim',
 'OpenCage',
 'OpenMapQuest',
 'Photon',
 'Point',
 'What3Words',
 'YahooPlaceFinder',
 'Yandex',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 'compat',
 'exc',
 'format',
 'geocoders',
 'get_geocoder_for_service',
 'location',
 'point',
 'units',
 'util']

In [51]:
type(df_geopy)

pandas.core.frame.DataFrame

In [52]:
df_geopy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103 entries, 0 to 102
Data columns (total 4 columns):
PostalCode       103 non-null object
Borough          103 non-null object
Neighbourhood    103 non-null object
location         103 non-null object
dtypes: object(4)
memory usage: 4.0+ KB


### Import GeoPy:

In [53]:
pip install geopy

Note: you may need to restart the kernel to use updated packages.


In [54]:
from geopy.geocoders import Nominatim
print('Nominatim imported')

Nominatim imported


### Set connection to OpenStreeMap 

In [55]:
df_geopy['address']=df_geopy['PostalCode'] + ',' + df_geopy['Borough'] + ','+ df_geopy['Neighbourhood']
df_geopy.head()

PostalCode      Borough                         Neighbourhood  \
0        M1B  Scarborough                         Rouge,Malvern   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough       Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                Woburn   
4        M1H  Scarborough                             Cedarbrae   

                                            location  \
0                    M1B, Scarborough, Rouge,Malvern   
1  M1C, Scarborough, Highland Creek,Rouge Hill,Po...   
2  M1E, Scarborough, Guildwood,Morningside,West Hill   
3                           M1G, Scarborough, Woburn   
4                        M1H, Scarborough, Cedarbrae   

                                             address  
0                      M1B,Scarborough,Rouge,Malvern  
1  M1C,Scarborough,Highland Creek,Rouge Hill,Port...  
2    M1E,Scarborough,Guildwood,Morningside,West Hill  
3                             M1G,Scarborough,Woburn  
4                          M1H,Scarborough,Cedarbrae

In [56]:
nom = Nominatim()

In [57]:
n=nom.geocode('M1B, Scarborough, Rouge,Malvern')
n

Location(Malvern, Scarborough—Rouge Park, Scarborough, Toronto, Golden Horseshoe, Ontario, M1B 4Y7, Canada, (43.8091955, -79.2217008, 0.0))

In [58]:
n.latitude

43.8091955

In [59]:
type(n)

geopy.location.Location

### Watch out for None values

In [60]:
n2=nom.geocode('M1E Scarborough Guildwood,Morningside,West Hill')
print(n2)

None


In [61]:
n3=nom.geocode("Toronto, Morningside")
n3

Location(Morningside, Scarborough—Guildwood, Scarborough, Toronto, Ontario, M1G 3K5, Canada, (43.7826012, -79.2049579, 0.0))

#### As you can see the postalcodes are different!

In [62]:
n4=nom.geocode("Toronto, Morningside, M1E ")
print(n4)

None


In [63]:
n5=nom.geocode("Toronto, M1E ")
print(n5)

None


#### None again, ..., it might be better to just drop these entries :(

### Set the full address:

In [64]:
df_geopy['address'] = df_geopy['PostalCode'] + ', ' + df_geopy['Borough'] + ', ' +df_geopy['Neighbourhood']

In [65]:
df_geopy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103 entries, 0 to 102
Data columns (total 5 columns):
PostalCode       103 non-null object
Borough          103 non-null object
Neighbourhood    103 non-null object
location         103 non-null object
address          103 non-null object
dtypes: object(5)
memory usage: 4.8+ KB


In [66]:
df_geopy.head()

PostalCode      Borough                         Neighbourhood  \
0        M1B  Scarborough                         Rouge,Malvern   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough       Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                Woburn   
4        M1H  Scarborough                             Cedarbrae   

                                            location  \
0                    M1B, Scarborough, Rouge,Malvern   
1  M1C, Scarborough, Highland Creek,Rouge Hill,Po...   
2  M1E, Scarborough, Guildwood,Morningside,West Hill   
3                           M1G, Scarborough, Woburn   
4                        M1H, Scarborough, Cedarbrae   

                                             address  
0                    M1B, Scarborough, Rouge,Malvern  
1  M1C, Scarborough, Highland Creek,Rouge Hill,Po...  
2  M1E, Scarborough, Guildwood,Morningside,West Hill  
3                           M1G, Scarborough, Woburn  
4                        M1H, Scarborough, Cedarbrae

### Use 'address' to get geocode coordinates:

Geocoding (Latitude/Longitude Lookup) Required parameters in a geocoding request: address — The street address that you want to geocode, in the format used by the national postal service of the country concerned. Additional address elements such as business names and unit, suite or floor numbers should be avoided.

In [67]:
df_geopy['Coordinates'] =df_geopy['address'].apply(nom.geocode)
df_geopy.head()



PostalCode      Borough                         Neighbourhood  \
0        M1B  Scarborough                         Rouge,Malvern   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough       Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                Woburn   
4        M1H  Scarborough                             Cedarbrae   

                                            location  \
0                    M1B, Scarborough, Rouge,Malvern   
1  M1C, Scarborough, Highland Creek,Rouge Hill,Po...   
2  M1E, Scarborough, Guildwood,Morningside,West Hill   
3                           M1G, Scarborough, Woburn   
4                        M1H, Scarborough, Cedarbrae   

                                             address  \
0                    M1B, Scarborough, Rouge,Malvern   
1  M1C, Scarborough, Highland Creek,Rouge Hill,Po...   
2  M1E, Scarborough, Guildwood,Morningside,West Hill   
3                           M1G, Scarborough, Woburn   
4                        M1H, Scarborough, Cedarbrae   

                                         Coordinates  
0  (Malvern, Scarborough—Rouge Park, Scarborough,...  
1                                               None  
2                                               None  
3  (Woburn, Scarborough—Guildwood, Scarborough, T...  
4                                               None

### location object created at 'Coordinates', but there are only Coordinates 5 non-null objects! 


In [68]:
df_geopy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103 entries, 0 to 102
Data columns (total 6 columns):
PostalCode       103 non-null object
Borough          103 non-null object
Neighbourhood    103 non-null object
location         103 non-null object
address          103 non-null object
Coordinates      5 non-null object
dtypes: object(6)
memory usage: 5.6+ KB


In [69]:
df_geopy.Coordinates[0]

Location(Malvern, Scarborough—Rouge Park, Scarborough, Toronto, Golden Horseshoe, Ontario, M1B 4Y7, Canada, (43.8091955, -79.2217008, 0.0))

In [70]:
print(df_geopy.Coordinates[1])

None


In [71]:
df_geopy['latitude']=df_geopy['Coordinates'].apply(lambda x: x.latitude if x !=None else None)
df_geopy['longitude']=df_geopy['Coordinates'].apply(lambda x: x.longitude if x !=None else None)
df_geopy.head()

PostalCode      Borough                         Neighbourhood  \
0        M1B  Scarborough                         Rouge,Malvern   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough       Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                Woburn   
4        M1H  Scarborough                             Cedarbrae   

                                            location  \
0                    M1B, Scarborough, Rouge,Malvern   
1  M1C, Scarborough, Highland Creek,Rouge Hill,Po...   
2  M1E, Scarborough, Guildwood,Morningside,West Hill   
3                           M1G, Scarborough, Woburn   
4                        M1H, Scarborough, Cedarbrae   

                                             address  \
0                    M1B, Scarborough, Rouge,Malvern   
1  M1C, Scarborough, Highland Creek,Rouge Hill,Po...   
2  M1E, Scarborough, Guildwood,Morningside,West Hill   
3                           M1G, Scarborough, Woburn   
4                        M1H, Scarborough, Cedarbrae   

                                         Coordinates   latitude  longitude  
0  (Malvern, Scarborough—Rouge Park, Scarborough,...  43.809196 -79.221701  
1                                               None        NaN        NaN  
2                                               None        NaN        NaN  
3  (Woburn, Scarborough—Guildwood, Scarborough, T...  43.759824 -79.225291  
4                                               None        NaN        NaN

In [72]:
df_geopy.to_csv('geo_loc_py.csv')

## As just 5 addresses were fruitful, we will go on to use the given geo-location data.

In [73]:
print('The latitude of', df_geopy.address[0],  'is', df_geopy.latitude[0], 'and its longitude is',df_geopy.longitude[0])

The latitude of M1B, Scarborough, Rouge,Malvern is 43.8091955 and its longitude is -79.2217008


#### Now let's merge the given dataframes:

In [74]:
# Load the Pandas libraries with alias 'pd' 
import pandas as pd 
# Read data from file 'filename.csv' 
# (in the same directory that your python process is based)
# Control delimiters, rows, column names with read_csv (see later) 
data2 = pd.read_csv("geopy.csv") 
# Preview the first 5 lines of the loaded data 
data2.head()

Unnamed: 0 PostalCode      Borough                         Neighbourhood  \
0           0        M1B  Scarborough                         Rouge,Malvern   
1           1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union   
2           2        M1E  Scarborough       Guildwood,Morningside,West Hill   
3           3        M1G  Scarborough                                Woburn   
4           4        M1H  Scarborough                             Cedarbrae   

                                            location  
0                    M1B, Scarborough, Rouge,Malvern  
1  M1C, Scarborough, Highland Creek,Rouge Hill,Po...  
2  M1E, Scarborough, Guildwood,Morningside,West Hill  
3                           M1G, Scarborough, Woburn  
4                        M1H, Scarborough, Cedarbrae

In [75]:
data3 = pd.read_csv("Geospatial_Coordinates.csv") 
# Preview the first 5 lines of the loaded data 
data3.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

- Rename 'Postal Code'

In [76]:
data3.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)
data3.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [85]:
data1 = pd.merge(data3, data2, how='inner', on=None, left_on=None, right_on=None,
         left_index=False, right_index=False, sort=True,
         suffixes=('_x', '_y'), copy=True, indicator=False,
         validate=None)

data1.head()

PostalCode   Latitude  Longitude  Unnamed: 0      Borough  \
0        M1B  43.806686 -79.194353           0  Scarborough   
1        M1C  43.784535 -79.160497           1  Scarborough   
2        M1E  43.763573 -79.188711           2  Scarborough   
3        M1G  43.770992 -79.216917           3  Scarborough   
4        M1H  43.773136 -79.239476           4  Scarborough   

                          Neighbourhood  \
0                         Rouge,Malvern   
1  Highland Creek,Rouge Hill,Port Union   
2       Guildwood,Morningside,West Hill   
3                                Woburn   
4                             Cedarbrae   

                                            location  
0                    M1B, Scarborough, Rouge,Malvern  
1  M1C, Scarborough, Highland Creek,Rouge Hill,Po...  
2  M1E, Scarborough, Guildwood,Morningside,West Hill  
3                           M1G, Scarborough, Woburn  
4                        M1H, Scarborough, Cedarbrae

In [86]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103 entries, 0 to 102
Data columns (total 7 columns):
PostalCode       103 non-null object
Latitude         103 non-null float64
Longitude        103 non-null float64
Unnamed: 0       103 non-null int64
Borough          103 non-null object
Neighbourhood    103 non-null object
location         103 non-null object
dtypes: float64(2), int64(1), object(4)
memory usage: 6.4+ KB


- Rearrange columns and drop foreign key:

In [87]:
cols = data1.columns.tolist()
cols

['PostalCode',
 'Latitude',
 'Longitude',
 'Unnamed: 0',
 'Borough',
 'Neighbourhood',
 'location']

In [88]:
new_column_order = ['PostalCode',
 'Borough',
 'Neighbourhood',
 'Latitude',
 'Longitude']
new_column_order

['PostalCode', 'Borough', 'Neighbourhood', 'Latitude', 'Longitude']

In [110]:
data1 = data1[new_column_order]
data1.head()

PostalCode      Borough                         Neighbourhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

- Sort dataframe to match example:

In [127]:
sorted_df = data1.sort_values([ 'Neighbourhood', 'Latitude'], ascending=[True, True])
sorted_df.head()
# no idea how to get it exacly like the exqample :(

PostalCode          Borough  \
58         M5H  DowntownToronto   
12         M1S      Scarborough   
14         M1V      Scarborough   
101        M9V        Etobicoke   
89         M8W        Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
58                              Adelaide,King,Richmond  43.650571 -79.384568  
12                                           Agincourt  43.794200 -79.262029  
14   Agincourt North,L'Amoreaux East,Milliken,Steel...  43.815252 -79.284577  
101  Albion Gardens,Beaumond Heights,Humbergate,Jam...  43.739416 -79.588437  
89                               Alderwood,Long Branch  43.602414 -79.543484

In [130]:
sorted_df.reset_index(inplace=True)
sorted_df.head()


level_0  index PostalCode          Borough  \
0        0     58        M5H  DowntownToronto   
1        1     12        M1S      Scarborough   
2        2     14        M1V      Scarborough   
3        3    101        M9V        Etobicoke   
4        4     89        M8W        Etobicoke   

                                       Neighbourhood   Latitude  Longitude  
0                             Adelaide,King,Richmond  43.650571 -79.384568  
1                                          Agincourt  43.794200 -79.262029  
2  Agincourt North,L'Amoreaux East,Milliken,Steel...  43.815252 -79.284577  
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...  43.739416 -79.588437  
4                              Alderwood,Long Branch  43.602414 -79.543484

In [131]:
sorted_cols =sorted_df.columns.tolist()
sorted_cols

['level_0',
 'index',
 'PostalCode',
 'Borough',
 'Neighbourhood',
 'Latitude',
 'Longitude']

In [132]:
new_column_order2 = ['PostalCode',
 'Borough',
 'Neighbourhood',
 'Latitude',
 'Longitude']
new_column_order2

['PostalCode', 'Borough', 'Neighbourhood', 'Latitude', 'Longitude']

In [133]:
sorted_dataframe = sorted_df[new_column_order]
sorted_dataframe.head()

PostalCode          Borough  \
0        M5H  DowntownToronto   
1        M1S      Scarborough   
2        M1V      Scarborough   
3        M9V        Etobicoke   
4        M8W        Etobicoke   

                                       Neighbourhood   Latitude  Longitude  
0                             Adelaide,King,Richmond  43.650571 -79.384568  
1                                          Agincourt  43.794200 -79.262029  
2  Agincourt North,L'Amoreaux East,Milliken,Steel...  43.815252 -79.284577  
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...  43.739416 -79.588437  
4                              Alderwood,Long Branch  43.602414 -79.543484

## 2.6. Submit a link to your Notebook on your Github repository. (2 marks)

In [134]:
sorted_dataframe.to_csv('sorted_geoloc.csv')

This notebook is an assignment for a course on **Coursera** called *Applied Data Science Capstone*, you can take this course online by clicking [here](http://cocl.us/DP0701EN_Coursera_Week3_LAB2).